{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load in needed mods\n",
    "import pandas as pd, numpy as np, glob, hypermutatorqtl as hypf\n",
    "from Bio.Seq import Seq\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Isvar</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Type</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Zeropos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1042367</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1042368</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1042369</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1042370</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1042371</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Pos Ref Alt  Isvar  Strand  Type  Phase              Gene  Sample  \\\n",
       "0  1042367   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1  1042368   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "2  1042369   G   G      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "3  1042370   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "4  1042371   C   C      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "\n",
       "   Zeropos  \n",
       "0        1  \n",
       "1        2  \n",
       "2        3  \n",
       "3        4  \n",
       "4        5  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Set path to long COA and load\n",
    "longCOApath = '../GENOTYPE/GENES/P19/CNAG_01836-t26_1.csv.gz'\n",
    "longCOA = pd.read_csv(longCOApath)\n",
    "\n",
    "## If negatively oriented, complement and flip\n",
    "if longCOA.Strand.max()<0:\n",
    "    longCOA = longCOA[::-1]\n",
    "    longCOA['Ref'] = [Seq(s).complement()[0] for s in longCOA.Ref.tolist()]\n",
    "    longCOA['Alt'] = [Seq(s).complement()[0] for s in longCOA.Alt.tolist()]\n",
    "\n",
    "## Add zero position\n",
    "longCOA['Zeropos'] = np.abs(longCOA.Pos - longCOA.Pos.tolist()[0])+1\n",
    "\n",
    "## reset variant column\n",
    "longCOA['Isvar'] = 0\n",
    "longCOA.loc[(longCOA.Ref!=longCOA.Alt),'Isvar']=1\n",
    "\n",
    "## View head\n",
    "longCOA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Isvar</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Type</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Zeropos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1042367</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1042368</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1042369</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1042370</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1042371</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Pos Ref Alt  Isvar  Strand  Type  Phase              Gene  Sample  \\\n",
       "0  1042367   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1  1042368   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "2  1042369   G   G      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "3  1042370   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "4  1042371   C   C      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "\n",
       "   Zeropos  \n",
       "0        1  \n",
       "1        2  \n",
       "2        3  \n",
       "3        4  \n",
       "4        5  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## View start codon\n",
    "longCOA[(longCOA.Type==0)].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Isvar</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Type</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Zeropos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1891</th>\n",
       "      <td>1044258</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1892</th>\n",
       "      <td>1044259</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1893</th>\n",
       "      <td>1044260</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1894</th>\n",
       "      <td>1044261</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1895</th>\n",
       "      <td>1044262</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1896</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Pos Ref Alt  Isvar  Strand  Type  Phase              Gene  Sample  \\\n",
       "1891  1044258   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1892  1044259   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1893  1044260   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1894  1044261   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1895  1044262   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "\n",
       "      Zeropos  \n",
       "1891     1892  \n",
       "1892     1893  \n",
       "1893     1894  \n",
       "1894     1895  \n",
       "1895     1896  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## View stop-codon\n",
    "longCOA[(longCOA.Type==0)].tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1042367, 1044262)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Where is this gene on the chromosome?\n",
    "longCOA.Pos.min(),longCOA.Pos.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 522)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Gather reference allele (H99)\n",
    "ref = Seq(''.join(longCOA[(longCOA.Type==0) & (longCOA.Ref!='.')].Ref.tolist()))\n",
    "\n",
    "## Translate the allele \n",
    "refaa = ref.translate()\n",
    "\n",
    "## Count stop-codons and print length of ref\n",
    "list(refaa).count('*'),len(refaa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2, 522)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Gather alternative allele (Bt65)\n",
    "alt = Seq(''.join(longCOA[(longCOA.Type==0) & (longCOA.Alt!='.')].Alt.tolist()))\n",
    "\n",
    "## Translate\n",
    "altaa = alt.translate()\n",
    "\n",
    "## Count stop-codons and print length\n",
    "list(altaa).count('*'),len(altaa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "503"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Find this index of the non-sense mutation\n",
    "difix = altaa.index('*')\n",
    "difix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(Seq('FPKWIV'), Seq('FPK*IV'))"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## View the amino acids around the above index\n",
    "## Set the number of codons around it\n",
    "na = 3\n",
    "refaa[difix-na:difix+na],altaa[difix-na:difix+na]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## slice coding regions\n",
    "cds = longCOA[(longCOA.Type==0)].copy()\n",
    "\n",
    "## Add ref and alt amino acids\n",
    "cds['Refaa'] = np.concatenate([np.repeat(a,3) for a in refaa])\n",
    "cds['Altaa'] = np.concatenate([np.repeat(a,3) for a in altaa])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1840"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Find the relative position of new stop codon\n",
    "zp = cds[(cds.Altaa=='*')].Zeropos.min()\n",
    "zp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pos</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Isvar</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Type</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Zeropos</th>\n",
       "      <th>Refaa</th>\n",
       "      <th>Altaa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1830</th>\n",
       "      <td>1044197</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1831</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1831</th>\n",
       "      <td>1044198</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1832</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1832</th>\n",
       "      <td>1044199</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1833</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1833</th>\n",
       "      <td>1044200</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1834</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1834</th>\n",
       "      <td>1044201</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1835</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1835</th>\n",
       "      <td>1044202</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1836</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1836</th>\n",
       "      <td>1044203</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1837</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1837</th>\n",
       "      <td>1044204</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1838</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1838</th>\n",
       "      <td>1044205</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1839</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1839</th>\n",
       "      <td>1044206</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1840</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1840</th>\n",
       "      <td>1044207</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1841</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1841</th>\n",
       "      <td>1044208</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1842</td>\n",
       "      <td>W</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1842</th>\n",
       "      <td>1044209</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1843</td>\n",
       "      <td>I</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1843</th>\n",
       "      <td>1044210</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1844</td>\n",
       "      <td>I</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1844</th>\n",
       "      <td>1044211</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1845</td>\n",
       "      <td>I</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1845</th>\n",
       "      <td>1044212</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1846</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1846</th>\n",
       "      <td>1044213</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1847</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847</th>\n",
       "      <td>1044214</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNAG_01836-t26_1</td>\n",
       "      <td>19</td>\n",
       "      <td>1848</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Pos Ref Alt  Isvar  Strand  Type  Phase              Gene  Sample  \\\n",
       "1830  1044197   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1831  1044198   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1832  1044199   C   C      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1833  1044200   C   C      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1834  1044201   C   C      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1835  1044202   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1836  1044203   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1837  1044204   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1838  1044205   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1839  1044206   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1840  1044207   G   G      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1841  1044208   G   A      1       1     0      0  CNAG_01836-t26_1      19   \n",
       "1842  1044209   A   A      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1843  1044210   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1844  1044211   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1845  1044212   G   G      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1846  1044213   T   T      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "1847  1044214   G   G      0       1     0      0  CNAG_01836-t26_1      19   \n",
       "\n",
       "      Zeropos Refaa Altaa  \n",
       "1830     1831     F     F  \n",
       "1831     1832     F     F  \n",
       "1832     1833     F     F  \n",
       "1833     1834     P     P  \n",
       "1834     1835     P     P  \n",
       "1835     1836     P     P  \n",
       "1836     1837     K     K  \n",
       "1837     1838     K     K  \n",
       "1838     1839     K     K  \n",
       "1839     1840     W     *  \n",
       "1840     1841     W     *  \n",
       "1841     1842     W     *  \n",
       "1842     1843     I     I  \n",
       "1843     1844     I     I  \n",
       "1844     1845     I     I  \n",
       "1845     1846     V     V  \n",
       "1846     1847     V     V  \n",
       "1847     1848     V     V  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Mannually set the codons to plot\n",
    "to_plot = cds[(cds.Zeropos>=zp-(3*3)) & (cds.Zeropos<zp+(3*3))]\n",
    "\n",
    "## View dataframe\n",
    "to_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../DATA/AA_color_df.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-283a4073087e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m## Load in amino acid color map\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0maadf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../DATA/AA_color_df.csv'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mindex_col\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m## add in a color for k\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0maadf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maadf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'K'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'tab:blue'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    684\u001b[0m     )\n\u001b[1;32m    685\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 686\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    687\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    688\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    450\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    451\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 452\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfp_or_buf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    453\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    454\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    944\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    945\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 946\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    947\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    948\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1176\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"c\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1177\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"c\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1178\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1179\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1180\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"python\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   2006\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"usecols\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2007\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2008\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2009\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2010\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../DATA/AA_color_df.csv'"
     ]
    }
   ],
   "source": [
    "## Load in amino acid color map\n",
    "aadf = pd.read_csv('../DATA/AA_color_df.csv',index_col=0).T\n",
    "\n",
    "## add in a color for k\n",
    "aadf[aadf.T.index.values[-1]+1] = ['K','tab:blue']\n",
    "aadf = aadf.T\n",
    "\n",
    "## view tail\n",
    "aadf.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Set figure\n",
    "fig, ax = plt.subplots(1,1,figsize=(10,3))\n",
    "fig.set_facecolor('w')\n",
    "\n",
    "## Plot gene body\n",
    "plt.hlines(0,longCOA.Zeropos.min(),longCOA.Zeropos.max(),color='grey',linewidth=6);\n",
    "\n",
    "## Plot exons\n",
    "for i,j in cds.iterrows():\n",
    "    plt.vlines(j.Zeropos,-1,1,color='grey',alpha=1,linewidth=1)\n",
    "    \n",
    "## Plot variants  \n",
    "## Variants not in exons\n",
    "gvs = longCOA[(longCOA.Isvar==1) & (longCOA.Type!=0)]\n",
    "for i,j in gvs.iterrows():\n",
    "    if j.Ref!=j.Alt:\n",
    "        plt.vlines(j.Zeropos,0.35,-0.25,color='k',linewidth=1)\n",
    "\n",
    "## Variant in exons\n",
    "gvs = longCOA[(longCOA.Isvar==1) & (longCOA.Type==0)]\n",
    "for i,j in gvs.iterrows():\n",
    "    if j.Ref!=j.Alt:\n",
    "        plt.vlines(j.Zeropos,1,-1,color='k',linewidth=1)\n",
    "        \n",
    "## Plot sequences\n",
    "## Set plotting variables for annotation of amino acids\n",
    "x = 1100\n",
    "ys = 0.7\n",
    "y1 = 8\n",
    "y2 = y1+2.5\n",
    "tw = 45\n",
    "xpad = 20\n",
    "\n",
    "## Initialize lists\n",
    "plotpos = []\n",
    "difpos = []\n",
    "\n",
    "## Iterate over codons and sequence to plot\n",
    "for i,(ri,p) in enumerate(to_plot.iterrows()):\n",
    "    \n",
    "    ## Set the x postion and append\n",
    "    xp = x+(i*tw)\n",
    "    plotpos.append(xp)\n",
    "    \n",
    "    ## Annotate the reference allele  \n",
    "    plt.text(xp,y=y1+ys,s=Seq(p.Ref),fontsize=12,color='k',va='top',ha='center')\n",
    "    \n",
    "    ## Annotate the alternatve allele\n",
    "    plt.text(xp,y=y2+ys,s=Seq(p.Alt),fontsize=12,color='k',va='bottom',ha='center')\n",
    "    \n",
    "    ## draw line connecting alleles if equal, else use a dot for variant\n",
    "    plt.text(xp,y=(y1+((y2-y1)/2))+ys+0.02,s='|' if p.Alt==p.Ref else '.',fontsize=10,color='k',va='center',ha='center')\n",
    "    \n",
    "    ## if allele don't match append dif position\n",
    "    if p.Alt!=p.Ref:\n",
    "        difpos.append(xp)\n",
    "    \n",
    "## Plot the reference amino acid sequence\n",
    "for i,a in enumerate(to_plot.Refaa.tolist()[::3]):  \n",
    "    \n",
    "    ## draw the rectagle representing the amino acid\n",
    "    plt.hlines(y1-1.6,plotpos[::3][i]-xpad,plotpos[2::3][i]+xpad,linewidth=15,alpha=0.75,color=aadf[(aadf.AA==a)].Color)\n",
    "\n",
    "    ## Annotate the amino acid\n",
    "    plt.text(plotpos[1::3][i],y=y1-1.6,s=a,color='w',va='center',ha='center')\n",
    "\n",
    "## Plot the alternative amino acid sequence\n",
    "for i,a in enumerate(to_plot.Altaa.tolist()[::3]):\n",
    "\n",
    "    ## draw the rectagle representing the amino acid\n",
    "    plt.hlines(y2+3.4,plotpos[::3][i]-xpad,plotpos[2::3][i]+xpad,linewidth=15,alpha=0.75,color=aadf[(aadf.AA==a)].Color)\n",
    "    \n",
    "    ## annotate the amino acid\n",
    "    plt.text(plotpos[1::3][i],y=y2+3.4,s=a if a != '*' else 'STOP',color='w',va='center',ha='center')\n",
    "\n",
    "## Gather xticks    \n",
    "x_xp,x_xl = plt.xticks()\n",
    "\n",
    "## Gather gene position values\n",
    "pos = longCOA[(longCOA.Zeropos.isin(x_xp[(x_xp>=500) & (x_xp<=longCOA.Zeropos.max())]))].Pos.values\n",
    "\n",
    "## Set relative position values\n",
    "zeropos = longCOA[(longCOA.Zeropos.isin(x_xp[(x_xp>=500) & (x_xp<=longCOA.Zeropos.max())]))].Zeropos.values\n",
    "\n",
    "## gather a new x positon modification\n",
    "newxmod = int(str(pos[0])[-2:])\n",
    "\n",
    "## addjust xticks and set x-axis lables\n",
    "plt.xticks(zeropos-newxmod , pos-newxmod ,fontsize=12)\n",
    "plt.xlabel('Chromosome 11 Coordinates (bp)',fontsize=12)\n",
    "\n",
    "## Annotate start and stop codons\n",
    "## The start\n",
    "plt.text(cds.Zeropos.min()+1,2.5,'ATG\\n%s'%cds.Zeropos.min(),va='center',ha='center',fontsize=12)\n",
    "\n",
    "## The stop\n",
    "plt.text(cds.Zeropos.max()-1,2.5,'TAA\\n%s'%str(cds.Zeropos.max()-2),va='center',ha='center',fontsize=12)\n",
    "\n",
    "## Set y-axis limits\n",
    "plt.ylim(-3.5,15);\n",
    "\n",
    "## Turn off top, left and right spines\n",
    "[ax.spines[a].set_visible(False) for a in ['top','left','right']]\n",
    "\n",
    "## Remove yticks\n",
    "plt.yticks([])\n",
    "\n",
    "## Annotate H99 and Bt65 strain lables\n",
    "plt.text(x-2.5*tw,y1,'%s:'%hypf.h99_label, fontsize=14,va='top',ha='center')\n",
    "plt.text(x-2.5*tw,y2+1.5, hypf.bt_label+':', fontsize=14,va='bottom',ha='center')\n",
    "\n",
    "## Annotate nonsense mutation\n",
    "plt.text(cds[(cds.Altaa=='*')].Zeropos[:3].max(),-2,'%s'%cds[(cds.Altaa=='*')].Zeropos[:3].max(),va='center',ha='center',fontsize=12)\n",
    "\n",
    "## plot a line between nonsnese mutation and amino acid in H99\n",
    "plt.plot([cds[(cds.Altaa=='*')].Zeropos.min(),np.max(difpos)],[1.3,y1-3],color='k',alpha=0.5,linewidth=1);\n",
    "\n",
    "## Set title (or dont')\n",
    "#plt.title('longCOA',x=0.1,y=0.8,fontstyle='italic',color='grey',fontsize=14)\n",
    "\n",
    "## save figure\n",
    "plt.savefig('../FIGURES/longCOA_Bt65vsH99.pdf',dpi=400,bbox_inches='tight');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Check our work\n",
    "assert np.max([np.unique([len(a) for a in longCOA.Ref]),np.unique([len(a) for a in longCOA.Alt])])==1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
